pyvex/lifting/util/lifter_helper.py

```diff
@@ -86,21 +103,47 @@ def lift(self, disassemble=False, dump_irsb=False):
 
         if disassemble:
             return [instr.disassemble() for instr in instructions]
-        self.irsb.jumpkind = JumpKind.Invalid
-        irsb_c = IRSBCustomizer(self.irsb)
+
         l.debug("Decoding complete.")
+        self.irsb.jumpkind = None
         for i, instr in enumerate(instructions[:self.max_inst]):
+            # first, create a scratch IRSB so we can throw the changes away if lifting fails
+            next_irsb_part = pyvex.IRSB.empty_block(self.irsb.arch, self.irsb.addr)
+            irsb_c = IRSBCustomizer(next_irsb_part)
+
+            # try to do the lifting
+            # if the instruction requires more context, we stop
+            # decoding here and just return the block we have so far
             l.debug("Lifting instruction " + instr.name)
-            instr(irsb_c, instructions[:i], instructions[i+1:])
-            if irsb_c.irsb.jumpkind != JumpKind.Invalid:
+            try:
+                instr(irsb_c, instructions[:i], instructions[i+1:])
+            except RequireContextError:
+                break
+
+            # lifting was successful so add the part to the full IRSB
+            self.irsb.extend(next_irsb_part)
+
+            # if the irsb has a conditional exit, end the block here
+            # angr doesn't like blocks that have conditional jumps in the middle (not at the end)
+            if irsb_c.has_conditional_exit:
+                self.irsb.next = Const(vex_int_class(self.irsb.arch.bits)(self.irsb.addr + self.irsb.size))
+                self.irsb.jumpkind = JumpKind.Boring
                 break
-            elif (i+1) == self.max_inst: # if we are on our last iteration
-                instr.jump(None, irsb_c.irsb.addr + irsb_c.irsb.size)
+
+            # if the block has a jumpkind, this is a full block so we can exit 
+            if self.irsb.jumpkind is not None:
                 break
-        else:
-            if len(irsb_c.irsb.statements) == 0:
-                raise LiftingException('Could not decode any instructions')
-            irsb_c.irsb.jumpkind = JumpKind.NoDecode
+
+        if len(self.irsb.statements) == 0:
```

Comment:
Would `if not self.irsb.statements` work? Pylint should complain about `len(...) == 0`.